kubeca/certinit/certinit.go

```diff
@@ -0,0 +1,192 @@
+	"strings"
+
+	"github.com/go-phorce/dolly/xlog"
+	"github.com/juju/errors"
+	capi "k8s.io/api/certificates/v1"
+	v1 "k8s.io/api/core/v1"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+var logger = xlog.NewPackageLogger("github.com/ekspand/trusty", "certinit")
+
+// Request parameters
+type Request struct {
+	// Namespace as defined by pod.metadata.namespace
+	Namespace string
+	// PodName name as defined by pod.metadata.name
+	PodName string
+	// CertDir is directory where the TLS certs should be written
+	CertDir string
+	// ClusterDomain specifies kubernetes cluster domain
+	ClusterDomain string
+	// Labels to include in CertificateSigningRequest object; comma separated list of key=value
+	Labels string
+	// QueryK8s specifies to query kubernetes for names appropriate to this Pod
+	QueryK8s bool
+	// SAN is additional comma separated DNS, IP, URI or Emails to include in SAN
+	SAN string
+	// ServiceNames specifies additional service names that resolve to this Pod; comma separated
+	ServiceNames string
+	// IncludeUnqualified specifies to include unqualified .svc domains in names from --query-k8s
+	IncludeUnqualified bool
+	// SignerName specifies the signer name
+	SignerName string
+
+	san       []string
+	labelsMap map[string]string
+}
+
+// CertClient provides minimum interfaces to create a CSR
+type CertClient struct {
+	Pods         MinPods
+	Services     MinServices
+	Certificates MinCertificates
+}
+
+// MinPods is minimum Pods interface
+type MinPods interface {
+	Get(ctx context.Context, name string, opts metaV1.GetOptions) (*v1.Pod, error)
+}
+
+// MinServices is minimum Services interface
+type MinServices interface {
+	List(ctx context.Context, opts metaV1.ListOptions) (*v1.ServiceList, error)
+}
+
+// MinCertificates is minimum Certificates interface
+type MinCertificates interface {
+	Create(ctx context.Context, certificateSigningRequest *capi.CertificateSigningRequest, opts metaV1.CreateOptions) (*capi.CertificateSigningRequest, error)
+	Get(ctx context.Context, name string, opts metaV1.GetOptions) (*capi.CertificateSigningRequest, error)
+}
+
+// Create certificate request and wait for issuance
+func (r *Request) Create(ctx context.Context, client *CertClient) error {
+	if r.Namespace == "" {
+		return errors.New("missing required namespace parameter")
+	}
+	if r.PodName == "" {
+		return errors.New("missing required pod name parameter")
+	}
+	if r.SignerName == "" {
+		return errors.New("missing required signer name parameter")
+	}
+
+	logger.Infof("Request, ns=%s, pod=%s, signer=%s, san=%q",
+		r.Namespace, r.PodName, r.SignerName, r.SAN)
+
+	// Gather the list of labels that will be added to the CreateCertificateSigningRequest object
+	r.labelsMap = make(map[string]string)
+	for _, n := range strings.Split(r.Labels, ",") {
+		if n == "" {
+			continue
+		}
+		s := strings.Split(n, "=")
+		label, key := s[0], s[1]
```

Comment:
check if len(s) > 2 otherwise this will throw index out of range exception and panic.